Fix `GraphLinksConverter::serialize_to` data alignment issues

ffuugoo opened the pull request:
This PR fix data alignment issue in `GraphLinksConverter::serialize_to` that was discovered in #3802.

It also adds a few asserts to the `transmute_from_u8*` functions, to try to catch similar alignment mismatches in the future.

Fixes #3802. Includes #3807.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

ffuugoo commented:
@generall @timvisee @IvanPleshkov I think this asserts *should* catch alignment mismatches in our code? Let's see if it will break any tests.

IvanPleshkov commented:
@ffuugoo Do you know how to easily fix storage compatibility test? It passes but logs are full of errors. Log snippet:
```
2024-03-11T13:45:16.587922Z ERROR qdrant::startup: Panic occurred in file /home/runner/work/qdrant/qdrant/lib/common/memory/src/mmap_ops.rs at line 122: assertion `left == right` failed: transmuting byte slice 0x0x7f3143d0cbe4 into slice of u64: required alignment is 8 bytes, byte slice misaligned by 4 bytes
  left: 4
 right: 0    
2024-03-11T13:45:16.589293Z DEBUG reqwest::connect: starting new connection: https://staging-telemetry.qdrant.io/    
2024-03-11T13:45:16.981934Z ERROR collection::shards::replica_set: Failed to load local shard "./storage/collections/test_collection/0", initializing "dummy" shard instead: Service internal error: Can't join segment load thread: TypeId { t: 42966343538335219590177265727833432740 }    
```

timvisee reviewed (commented):
:+1:

ffuugoo commented:
> Do you know how to easily fix storage compatibility test?

Do you mean, "how to make these tests fail, instead of passing with a dummy"? If so, not sure.

This PR (so far) mostly to show that we *do* have (mis)alignment issues and that they are easy to reproduce. The best fix would be to fix the alignment issues in the first place.

ffuugoo commented on lib/segment/src/index/hnsw_index/graph_links.rs:
```diff
@@ -224,9 +255,10 @@ impl GraphLinksConverter {
             let offsets_range = header.get_offsets_range();
             let union_range = links_range.start..offsets_range.end;
-            let (links_mmap, offsets_mmap) = bytes_data[union_range]
+            let (links_mmap, offsets_with_padding_mmap) = bytes_data[union_range]
                 .as_mut()
                 .split_at_mut(links_range.len());
+            let offsets_mmap = &mut offsets_with_padding_mmap[header.offsets_padding as _..];
```
I've tweaked @IvanPleshkov's code here, to make sure `offsets_padding` betes *aren't* included as (an invalid) last element of `links_mmap`.

But, please, check my slicing here....

IvanPleshkov replied:
> I've tweaked @IvanPleshkov's code here, to make sure `offsets_padding` betes _aren't_ included as (an invalid) last element of `links_mmap`. (See [this comment](https://github.com/qdrant/qdrant/pull/3807#discussion_r1521806532) from #3807.)
> 
> But, please, check my slicing here....

Thanks for the fix, you are totally right and I like your opinion!

ffuugoo commented:
Thanks @IvanPleshkov for fixing the alignment issues in `GraphLinks*`!

I've updated the PR description, switched alignment asserts back to `debug_assert` and added a small fix. It's ready for the final review, I think.

agourlay commented:
Maybe we could merge first https://github.com/qdrant/qdrant/pull/3816 and rebase.
It introduces more coverage for mmap configurations.

agourlay commented:
Merged, please rebase :gear:
